src/Chapter3.hs

```diff
@@ -1006,6 +1149,50 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data BattleResult = Victory | Defeat
+
+class Fighter x y where
+    attackEm :: x -> y -> y
+    runAway :: x -> y -> BattleResult
+
+data Knight2 = Knight2
+    { k2Defence :: Int
+    , k2AttackPower :: Int
+    , k2Health :: Int
+    }
+
+data Monster2 = Monster2
+    { m2AttackPower :: Int
+    , m2Health :: Int
+    }
+
+instance Fighter Knight2 Monster2 where
+    attackEm :: Knight2 -> Monster2 -> Monster2
+    attackEm (Knight2 kd ka kh) (Monster2 ma mh) = Monster2 (ma) (mh - ka)
+    runAway :: Knight2 -> Monster2 -> BattleResult
+    runAway (Knight2 kd ka kh) (Monster2 ma mh) = Defeat
+
+instance Fighter Monster2 Knight2 where
+    attackEm :: Monster2 -> Knight2 -> Knight2
+    attackEm (Monster2 ma mh) (Knight2 kd ka kh) = Knight2 kd ka (kd + kh - ma)
+    runAway :: Monster2 -> Knight2 -> BattleResult
+    runAway (Monster2 ma mh) (Knight2 kd ka kh) = Victory
+
+-- Idk how to do this man
+
+-- roundOfFight :: BattleAction -> Monster2 -> Knight2 -> BattleResult
```

Comment:
Don't worry! 👍🏼 

You can add the constraint about fighters and then use the methods from the typeclass.
For example:

```suggestion
roundOfFight :: (Fighter m, Fighter k) => BattleAction -> m -> k -> BattleResult
```